src/main/java/in/projecteka/consentmanager/user/OtpRequestAttemptService.java

```diff
@@ -0,0 +1,76 @@
+package in.projecteka.consentmanager.user;
+
+import in.projecteka.consentmanager.clients.ClientError;
+import in.projecteka.consentmanager.user.model.OtpRequestAttempt;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Mono;
+
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+@AllArgsConstructor
+public class OtpRequestAttemptService {
+
+    private final OtpRequestAttemptRepository otpRequestAttemptRepository;
+    private final UserServiceProperties userServiceProperties;
+
+    public Mono<Void> validateOTPRequest(String identifierType, String identifierValue, OtpRequestAttempt.Action action, String cmId) {
+        return otpRequestAttemptRepository.getOtpAttempts(cmId, identifierType, identifierValue, userServiceProperties.getMaxOtpAttempts(), action)
+                .filter(otpAttempts -> otpAttempts.size() >= userServiceProperties.getMaxOtpAttempts())
+                .filter(this::isNoneBlockedExceptLatest)
+                .flatMap(this::validateLatestAttempt)
+                .flatMap(this::validateAttemptsLimit)
+                .switchIfEmpty(Mono.defer(() -> createOtpAttemptFor(cmId, identifierType, identifierValue, OtpRequestAttempt.AttemptStatus.SUCCESS, action)));
+    }
+
+    public Mono<Void> validateOTPRequest(String identifierType, String identifierValue, OtpRequestAttempt.Action action) {
+        return validateOTPRequest(identifierType, identifierValue, action, "");
+    }
+
+    private Mono<Void> createOtpAttemptFor(String cmId, String identifierType, String identifierValue, OtpRequestAttempt.AttemptStatus attemptStatus, OtpRequestAttempt.Action action) {
+        return otpRequestAttemptRepository.insert(cmId, identifierType, identifierValue, attemptStatus, action);
+    }
+
+    private Mono<List<OtpRequestAttempt>> validateLatestAttempt(List<OtpRequestAttempt> attempts) {
+        OtpRequestAttempt latestAttempt = attempts.get(0);
+        if (latestAttempt.getAttemptStatus() == OtpRequestAttempt.AttemptStatus.FAILURE) {
+            return handleBlockedLatestAttempt(latestAttempt);
+        }
+        return Mono.just(attempts);
+    }
+
+    private Mono<List<OtpRequestAttempt>> handleBlockedLatestAttempt(OtpRequestAttempt latestAttempt) {
```

Comment:
you can use flux here.

Reply:
Actually we need to have complete list to perform logic on it.

Reply:
Mono<List> is same as Flux.

Reply:
Yes. But there will be additional `flux.collectList()` we will have to in order to get the list, right? Please let me know if there is a way to collect list from flux without blocking.